src/drivers/safety_button/safety.cpp

```diff
@@ -0,0 +1,351 @@
+{
+	orb_unsubscribe(_armed_sub);
+
+	orb_unadvertise(_to_safety);
+}
+
+void
+SafetyButton::safety_check_button(void)
+{
+	if (!PX4_MFT_HW_SUPPORTED(PX4_MFT_PX4IO)) {
+
+		static int counter = 0;
+		/*
+		 * Debounce the safety button, change state if it has been held for long enough.
+		 *
+		 */
+		bool safety_button_pressed = px4_arch_gpioread(GPIO_BTN_SAFETY);
+
+		/*
+		 * Keep pressed for a while to arm.
+		 *
+		 * Note that the counting sequence has to be same length
+		 * for arming / disarming in order to end up as proper
+		 * state machine, keep ARM_COUNTER_THRESHOLD the same
+		 * length in all cases of the if/else struct below.
+		 */
+		if (safety_button_pressed && !_safety_btn_off) {
+
+			if (counter < CYCLE_COUNT) {
+				counter++;
+
+			} else if (counter == CYCLE_COUNT) {
+				/* switch to armed state */
+				_safety_btn_off = true;
+				counter++;
+			}
+
+		} else if (safety_button_pressed && _safety_btn_off) {
+
+			if (counter < CYCLE_COUNT) {
+				counter++;
+
+			} else if (counter == CYCLE_COUNT) {
+				/* change to disarmed state and notify the FMU */
+				_safety_btn_off = false;
+				counter++;
+			}
+
+		} else {
+			counter = 0;
+		}
+	}
+}
+
+void
+SafetyButton::flash_safety_button()
+{
+	if (!PX4_MFT_HW_SUPPORTED(PX4_MFT_PX4IO)) {
+
+		bool armed_updated = false;
+		orb_check(_armed_sub, &armed_updated);
+
+		if (armed_updated) {
+			orb_copy(ORB_ID(actuator_armed), _armed_sub, &_armed);
+		}
+
+		/* Select the appropriate LED flash pattern depending on the current arm state */
+		uint16_t pattern = LED_PATTERN_FMU_REFUSE_TO_ARM;
+
+		/* cycle the blink state machine at 10Hz */
+		static int blink_counter = 0;
+
+		if (_safety_btn_off) {
+			if (_armed.armed) {
+				pattern = LED_PATTERN_IO_FMU_ARMED;
+
+			} else {
+				pattern = LED_PATTERN_IO_ARMED;
+			}
+
+		} else if (_armed.armed) {
+			pattern = LED_PATTERN_FMU_ARMED;
+
+		} else {
+			pattern = LED_PATTERN_FMU_OK_TO_ARM;
+
+		}
+
+		/* Turn the LED on if we have a 1 at the current bit position */
+		px4_arch_gpiowrite(GPIO_LED_SAFETY, !(pattern & (1 << blink_counter++)));
+
+		if (blink_counter > 15) {
+			blink_counter = 0;
+		}
+	}
+}
+
+int
+SafetyButton::task_spawn(int argc, char *argv[])
+{
+	/* schedule a cycle to start things */
+	int ret = work_queue(HPWORK, &_work, (worker_t)&SafetyButton::cycle_trampoline, nullptr, 0);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	_task_id = task_id_is_work_queue;
+
+	return PX4_OK;
+}
+
+void
+SafetyButton::cycle_trampoline(void *arg)
+{
+	SafetyButton *dev = reinterpret_cast<SafetyButton *>(arg);
+
+	// check if the trampoline is called for the first time
+	if (!dev) {
+		dev = new SafetyButton();
+
+		if (!dev) {
+			PX4_ERR("alloc failed");
+			return;
+		}
+
+		_object = dev;
+	}
+
+	dev->cycle();
+}
+
+void
+SafetyButton::run()
+{
+	cycle();
+}
+
+void
+SafetyButton::cycle()
+{
+	const hrt_abstime cycle_timestamp = hrt_absolute_time();
+
+	if (!PX4_MFT_HW_SUPPORTED(PX4_MFT_PX4IO)) {
```

Comment:
Instead of checking here every time, can we avoid starting the module in the first place?